derivatives of self energies

wkotlarski opened the pull request:
This PR introduces derivatives of self energies in autogenerated models. @InflammableSausage computed self energies and their derivatives in the SM using FeynArts/FormCalc and prepared a unit test.

TODO:

- [ ] decays fail in semianalytic models with: `CMSSMSemiAnalytic_ewsb_solver<Semi_analytic>:solve: pointer to semi-analytic solutions is zero!`. This is because the solver doesn't operate on the interface. Revert c5bbbf2045a9466b0a1963b2224e4efafea70ba2 after the fix.

wkotlarski committed:
fixed comment

wkotlarski committed:
dummy softsusy functions

Commit:
simplify softsusy dummy functions

Commit:
fflite dummy functions

Commit:
tiny tweaks

wkotlarski committed:
single call DB functions work

Commit:
call only single DB functions

Commit:
test derivatives

Commit:
test derivatives only with LT or Collier

Commit:
added derivatives of H0, F0 and G0

wkotlarski committed:
Merge branch 'development' into feature-self-energies-derivatives

wkotlarski committed:
Merge branch 'development' into feature-self-energies-derivatives

Commit:
collier: all libraries

Commit:
looptools: all libraries

Commit:
fixed comment

Commit:
softsusy DB0

Commit:
remove uneeded variables

Commit:
make stuff const

Commit:
Revert "remove uneeded variables"

This reverts commit f929ae899515adf104bda2d6637159aa5c69a657.

Commit:
Merge remote-tracking branch 'origin/feature-self-energies-derivatives' into feature-self-energies-derivatives2

Commit:
added comment

Commit:
added comment

Commit:
Merge remote-tracking branch 'origin/feature-self-energies-derivatives' into feature-self-energies-derivatives2

Commit:
Merge remote-tracking branch 'origin/development' into feature-self-energies-derivatives2

Commit:
rename local variable

p conflicts with p in expression for self energies

Commit:
rever unneeded change

wkotlarski commented:
@Expander, should we generate self energy derivatives for all particles or only the ones that are strictly required?

Expander commented:
Hmm, good question. I guess it would be easier for all particles we have, right?

wkotlarski committed:
add derivatives of loop functions

wkotlarski commented:
It would be easier to generate derivatives for all self energies. But we can implement the filter later. 

The other issue is that we'll also need self energies and derivatives in the decoupling model. More precisely, we are gonna call then via the `cxx_qft` stuff from Jobst, which operates on `mass_eigenstates_interface`. Would you be ok with moving self energies, derivatives and other required things to the interface class. We'll also need to create a wrapper in `cxx_qft` class that will internally call those functions. Something like
```cpp
template<> inline
std::complex<double> context_base::self_energy_1loop_deriv_p2<fields::Fe>(double p) const
{ return model.self_energy_Fe_1loop_deriv_p2(p); }
```
Or maybe you have a better idea. I'm also open to suggestions.

wkotlarski committed:
Merge branch 'development' into feature-self-energies-derivatives2

Expander commented:
After our discussion, I've noticed that (besides the vertices and self-energies) one would also move the masses and mixing matrices and the corresponding helper functions for the mass matrices to the new `<model_name>_self_energy` class.

wkotlarski commented:
Are you sure? Then what would left in `<model_name>_mass_eigenstates`?  It would be just the initialization of parameters? Should we then rename  `<model_name>_mass_eigenstates` to something else?

Expander commented:
I'm pretty sure, because the self-energies require the masses in the loop functions. But you're asking the right question. It seems to me that there is nothing left in the old `mass_eigenstates` class and it would make sense to rename it. 

Here is a new proposal:

1. We restore the (old) inheritance hierarchy of the non-template classes: `<model>_susy_parameters` <- `<model>_soft_parameters` <- `<model>_mass_eigenstates`, as in FS 1.0. Here `<model>_mass_eigenstates` is basically the current "mass eigenstates" class (it contains the masses, mixing matrices and self-energies).

1. We create a new "non-decoupling mass eigenstates class", let's call it `<model>_mass_eigenstates_running` for the moment. It is derived from
   * `<model>_mass_eigenstates_interface`
   * `<model>_mass_eigenstates`

1. The "decoupling mass eigenstates class" `<model>_mass_eigenstates_decoupling` is derived from

   * `<model>_mass_eigenstates_interface`
   * `<model>_mass_eigenstates`

I've sketched the proposal here: [Inheritance-proposal-2024-11-15.pdf](https://github.com/user-attachments/files/17778745/Inheritance-proposal-2024-11-15.pdf)

wkotlarski committed:
new file structure

wkotlarski commented:
Expander, what should `<model>_mass_eigenstates_running` and `<model>_mass_eigenstates_decoupling` contain? Because it seems that the content of them both is almost identical. So almost everything could be implemented in `<model>_mass_eigenstates`. It seems that `<model>_mass_eigenstates_decoupling` will only contain the `fill` function.

wkotlarski commented:
In this setup, what would be left for `<model>_mass_eigenstates_running`? Also, can pure virtual functions from `<model>_mass_eigenstates_interface` be implemented via inheritance from the `<model>_mass_eigenstates` class or do they have to be implemented in `<model>_mass_eigenstates_decoupling` and `<model>_mass_eigenstates_running`?

wkotlarski committed:
update CI to Himalaya 4.2.3

wkotlarski commented:
All observables, like g-2 for example, use `mass_eigenstates`. So they'd have to be changed to `mass_eigenstates_running`, right?

Expander commented:
> All observables, like g-2 for example, use `mass_eigenstates`. So they'd have to be changed to `mass_eigenstates_running`, right? Or `weiberg_angle`, etc.

Hmm, are you sure? Since `mass_eigenstates_running` is basically "empty", why can't they still operate on the `mass_eigenstates` class?

wkotlarski commented:
So would the `mass_eigenstates_running` be only used for thnigs that allow for switching between decoupling and non-decoupling schemes? Which at this point are only the decays?

Expander commented:
Yes, I think it would be like this.

Of course, `mass_eigenstates_running` would still appear at some intermediate class in the class hierarchy, starting from `<model>_susy_parameters` and ending at `<model><Two_scale>`.

wkotlarski commented:
This
```mathematica
double get_g1() const override { return SUPER(g1); }
```
cannot be make by `MakeOverride`. Right? Because it doesn't add `SUPER`.

Expander commented:
I think `MakeOverride` does not add/remove content from the functions' body, as far as I can see.

wkotlarski commented:
If I understand correctly, `mass_eigenstates_interface` has a whole bunch of getters and setters and they'd all have to be implemented via `SUPER`

Expander commented:
Yes, that's true. :( Is it possible to adapt (or get inspired by) the code that generates the getters/setters that exist now to implement the `mass_eigenstates_interface`?

wkotlarski commented:
An obvious solution would be if all of getter/setter creating functions, e.g. `CreateModelParameterGetter`, had an extra option like `Super->False`. Otherwise I have to start parsing strings generated by functions like  `CreateModelParameterGetter` which is prone to errors.

Expander commented:
Arrg, parsing strings would be very bad. An extra option `super_:False` would be much better.

wkotlarski commented:
If a variable is private, it cannot be accessed from the derived class, right? For example, this doesn't seem to work
```cpp
double get_MVG() const override { return SM_mass_eigenstates::MVG; }
```

Expander commented:
Maybe it would be easier to make the member variables `protected` instead of `private`. (This is done in `<model>_soft_parameters` for example.) Then you can access them as
~~~.cpp
double get_MVG() const override { return SM_mass_eigenstates::MVG; }
~~~

wkotlarski commented:
I think I got most of the stuff working. The only missing thing are definitions (not declarations) of self energies and their derivatives in `<model>_mass_eigenstates_running.cpp` and `<model>_mass_eigenstates_decoupling_scheme.cpp`. Is there a function that can create them or should I write them myself. I guess they would be similar to `CreateModelParameterSetter`.

uukhas committed:
NPointFunctions tweaks (#632)

final changes from Vova

Commit:
add {} around for loop body

wkotlarski committed:
code compiles; self energies and derivatives are generated

wkotlarski commented:
Hi @Expander. What a mess. I got the code to the stage where self energies and derivatives are generated, the class structure we discussed is implemented and the code compiles. I have to clean in it and have a look at it again. Because I've lost control over all of those changes...

Commit:
wrapper to call self energies and ders. from cxx

wkotlarski committed:
fixed npointfunction compilation problem

Commit:
fixed FormCalc installation error on CI

to be reverted once it's fixed upstream by Hahn

wkotlarski committed:
more fixes

wkotlarski commented:
The effect of all of those changes is that every part of code that operates on things from Jobst now has to take `<model>_mass_eigenstates_running`. This includes pieces that compute g-2, edm, decays, NPointFunctions etc. 

In tests we have a lot of `setup_<model>` functions. For example, `setup_MRSSM2` returns `MRSSM2_slha` which cannot be convert to `MRSSM2_mass_eigenstates_running`
```
note: candidate function template not viable: no known conversion from 'const MRSSM2_slha' to 'const MRSSM2_mass_eigenstates_running' for 1st argument
```
Do you have a recommendation how to approach it? Should I tweak the `setup_<model>` or add some conversion from `<model>_mass_eigenstates_running` to `<model>_slha`?

wkotlarski committed:
loop functions are now protected, not private

Commit:
fixed CMSSM error

Commit:
fixed MRSSM2 test

Commit:
fixed some SM errors

Commit:
fixed SM test

Commit:
fixed CMSSMCKM test

Commit:
fixed more errors

wkotlarski committed:
fixed next SM test

Commit:
fixed MRSSM2CKM error

Commit:
fixed infinite loop

Commit:
reduced MRSSM2 test precision

Commit:
Merge remote-tracking branch 'origin/development' into feature-self-energies-derivatives2

Commit:
Revert "wrapper to call self energies and ders. from cxx"

This reverts commit 256a5827207a3be3e6db2ac0f011b2d3694d57d4.

wkotlarski commented:
I'm done. Here's the summary of changes:

- There were already variables that were suppose to contain derivatives and associated `@` tags,  like `selfEnergyDerivFunctions` , but they were empty since derivatives were contained together with self energies. I've removed them.
- Since now some functions are protected not private I had to change redefiniton of private to public in tests to protected to public.
- In all places that use Jobst's CXX stuff I've replaced `<model>_mass_eigenstates` with `<model>_mass_eigenstates_running`.
- `<model><Two_scale>` requires extra conversion. Had to adapt some tests, see `test_SM_FlexibleDecay.cpp` for example.
- `<model>_slha` doesn't convert automatically to `<model>_mass_eigenstates_running`. Had to change `<model>_slha` to  `<model>_mass_eigenstates_running` in some tests. See  for example `test_THDMII_FlexibleDecay.cpp`

Commit:
removed unneded if block

Expander commented:
Thanks a lot for the changes! I'll take a careful look at them within the next days.

Commit:
fixed segfault in calling self energies

Expander commented:
Sorry for the delay, I'm quite busy right now. I'll try to start looking at this PR by the end of the week.

wkotlarski commented:
No problem. That's not urgent. I can use it independently of whether it's in the development or not.

wkotlarski commented:
@Expander, there's a problem with this PR. In `SM_mass_eigenstates_decoupling_scheme.cpp` we call self energies from the base class
```cpp
std::complex<double> CLASSNAME::self_energy_hh_1loop(double p ) const {
   return SM_mass_eigenstates::self_energy_hh_1loop(p);
}
```
The problem with this is that if we print the tree-level masses from `SM_mass_eigenstates::self_energy_hh_1loop` they are 0. This seems to be because `Mhh` is also defined as private in `SM_mass_eigenstates_decoupling_scheme.hpp`. If I remove it, then masses are non-zero.

Expander commented:
Do you mean that if you make the masses `public`, they are non-zero? This is very strange and I cannot imagine that this is the root of the problem. (We can make the masses `protected`, if you like.)

Could it be that the masses are defined multiple times? And the modification of the access specifier (`private` or `public`) just changes which masses are filled?

wkotlarski commented:
I think you are right that the problem is that masses are defined twice: in the base (`<model>_mass_eigenstates`) and in the derived (`<model>_mass_eigenstates_decoupling_scheme`) class. We implement a pure virtual function by calling a function from one of base classes and it seems to access wrong masses. The solution that I've found is to remove masses from the derived class althogether.

Expander commented:
Yes, I also think that this is the best solution!

wkotlarski committed:
removed unneded members in a derived class

wkotlarski committed:
Merge remote-tracking branch 'origin/development' into feature-self-energies-derivatives2

Expander commented:
I've had a look at this first implementation and have a major comment:

In `<model>_mass_matrix_running` and `<model>_mass_matrix_decoupling` there are functions that return the mass matrices (`get_mass_matrix_Fu()`, ...) and solve the EWSB equations (`solve_ewsb_tree_level_custom()`, ...) that currently re-implement the expressions. I think these functions should instead return the corresponding value from the base class `<model>_mass_matrix`.

So far I've identified  the following functions that should probably not re-implement expressions but use expressions from the `<model>_mass_matrix` base class:

 print

 solve_ewsb_tree_level_custom
 ... // all EWSB functions

 calculate_tree_level_mass_spectrum
 calculate_pole_mass_spectrum
 copy_tree_level_masses_to_pole_masses
 calculate_mass_spectrum
 reorder_tree_level_masses

 clear_tree_level_parameters
 clear_problems
 clear

 set_input_parameters
 get_input_parameters
 set_extra_parameters
 get_extra_parameters

 get_mass_matrix_Fu
 get_mass_matrix_Fd
 ... // all mass matrix getters

 set_tree_level_masses
 get_tree_level_masses
 set_tree_level_masses_and_mixings
 get_tree_level_masses_and_mixings

wkotlarski commented:
Thanks a lot for looking into this unprompted! I'd like to bring this PR in next so was planning on going through this class hierarchy. I only briefly looked at your comment but it seems to align with what I've found. There's a problem with these overloads and sometimes, depending on how the object is accesses (object or pointer to an object), masses would be 0 or not. I'll have a look at it.

wkotlarski commented:
I think that because `get_mass_matrix_Fu` is defined as pure virtual in `<model>_mass_eigenstates_interface.hpp` it has to be implemented but doesn't have to be re-implemented. We could do the following
```cpp
Eigen::Matrix<double,3,3> <model>_mass_eigenstates_running::get_mass_matrix_Fu() {
   return <model>_mass_eigenstate::get_mass_matrix_Fu();
}

wkotlarski committed:
removed unneeded interface functions

Commit:
fixed error

wkotlarski committed:
fix

Commit:
fix missing define

wkotlarski commented:
@Expander, first, I've cleaned the `mass_eigenstates_interface`. I've removed a lot of the functions which were never used. Can you have a look at some point if you'd be ok with this PR?

wkotlarski committed:
Merge remote-tracking branch 'origin/development' into feature-self-energies-derivatives2

Commit:
fix

wkotlarski committed:
fixed test

wkotlarski committed:
fixed broken test

wkotlarski committed:
SM self-energy test

wkotlarski committed:
missing file

wkotlarski commented:
Hi @Expander. Sorry for all this mess. I think I've finally understood how this is suppose to work and fixed most of the problems. The only thing left is the issue with models using semi-analytic solver.

In decoupling models we solve EWSB in `fill_from` function. The call to base class `solve_ewsb_equations_tree_level` function works fine unless we use semi-analytic solver. I think this fail because `<model>_semi_analytic_ewsb_solver` works on `<model>__mass_eigenstates`. Do you have a suggestion how to restructure it?

wkotlarski committed:
Merge remote-tracking branch 'origin/development' into feature-self-energies-derivatives2

wkotlarski committed:
temporarily disable running CMSSMSemiAnalytic test

Expander committed:
use forward declaration where possible

in order to minimize compile time

Expander committed:
mark derived-class destructors as override

and = default if appropriate

Expander commented:
I think the `mass_eigenstates_interface` is pretty good now!

Expander committed:
fix SMSemiAnalytic test

Expander commented:
W.r.t. the failing semi-analytic solver: I can currently not see a failing test. Which failing test are you referring to?

Commit:
Revert "temporarily disable running CMSSMSemiAnalytic test"

This reverts commit c5bbbf2045a9466b0a1963b2224e4efafea70ba2.

wkotlarski commented:
Thanks for having a look! If you run a default `CMSSMSemiAnalytic` input file you'll get
```
Block DCINFO
     1   FlexibleSUSY
     2   2.8.0
     4   CMSSMSemiAnalytic_ewsb_solver<Semi_analytic>:solve: pointer to semi-analytic solutions is zero!
     5   CMSSMSemiAnalytic
     9   4.15.3
DECAY        25     0.00000000E+00   # hh(1) decays
DECAY        35     0.00000000E+00   # hh(2) decays
DECAY       -37     0.00000000E+00   # Hpm(2) decays
DECAY        36     0.00000000E+00   # Ah(2) decays
Block EFFHIGGSCOUPLINGS
```
The test that fails is `test_CMSSMSemiAnalytic_spectrum.sh`.

Expander committed:
remove FlexibleDecay block for comparison

Expander commented:
Ok, I can reproduce the issue now. (I had to enable the decays.)

Expander committed:
avoid spurious copy

Expander commented:
Ok, that's a tricky one. First, one question / suggestion for a simplification:

In `_decays.cpp` we're currently setting up the decoupling model as follows:
~~~.cpp
auto dm = std::make_unique<CMSSMSemiAnalytic_mass_eigenstates_decoupling_scheme>(model.get_input());
dm->fill_from(model);
~~~
I'd suggest to instead write
~~~.cpp
auto dm = std::make_unique<CMSSMSemiAnalytic_mass_eigenstates_decoupling_scheme>(model);
~~~
Then we could remove the `fill_from(CMSSMSemiAnalytic_mass_eigenstates)` function, which seems to be redundant to a constructor. Would that work?

Expander commented:
If it is ok for you, I'll try to make a small refactoring of the EWSB solver and EWSB solution classes. I'll do this in a separate branch and create a PR.

Expander commented:
If it is ok for you, I'll try to make a small refactoring of the EWSB solver and EWSB solution classes. I'll do this in a separate branch and create a PR.
